core/blockchain.go

```diff
@@ -2375,19 +2373,13 @@ func (bc *BlockChain) maintainTxIndex(ancients uint64) {
 
 // BadBlocks returns a list of the last 'bad blocks' that the client has seen on the network
 func (bc *BlockChain) BadBlocks() []*types.Block {
-	blocks := make([]*types.Block, 0, bc.badBlocks.Len())
-	for _, hash := range bc.badBlocks.Keys() {
-		if blk, exist := bc.badBlocks.Peek(hash); exist {
-			block := blk.(*types.Block)
-			blocks = append(blocks, block)
-		}
-	}
+	blocks, _ := rawdb.ReadAllBadBlocks(bc.db)
```

Comment:
Previously, the bad block were limited (an LRU IIRC?). With this change, a long-running node can potentially gather up hundreds (thousands?) of blocks, and if we try to serve all of them via RPC (full RLP including transactions), the call may fail where it previously would have succeeded. 

So if we change this, maybe we need to add an (optional) argument to `debug.getBadBlocks` to limit it. And/or add `debug.getBadBlock(<hash>)`

Reply:
Yes, agree.

Reply:
Added an optional arg `criteria` to `debug.getBadBlock` where you can specify the `from`, `to` and `target hashset`